middleware/gzip.go

```diff
@@ -41,265 +54,210 @@ func Gzip(wrappedHandler http.HandlerFunc) http.HandlerFunc {
 			return
 		}
 
-		gw := &gzipRW{
+		grw := &gzipRW{
 			ResponseWriter: w,
-			// h:              h,
-			buf: bufferPool.Get().(*[]byte),
+			// Note: do not set `gw` here, it will be set when `startGzip` is called.
+			// TODO: maybe we should set `gw` here??
+			level:             defaultLevel,
+			minSize:           defaultMinSize,
+			contentTypeFilter: defaultContentTypeFilter,
 		}
-		defer func() {
-			err := gw.Close()
-			if err != nil && os.Getenv("GOWEB_RUNNING_IN_TESTS") != "" {
-				panic(err)
-			}
-		}()
-
-		var modifiedRw http.ResponseWriter = gw
-
-		_, cok := w.(http.CloseNotifier)
-		_, hok := w.(http.Hijacker)
-		_, pok := w.(http.Pusher)
-
-		_ = cok
-		_ = hok
-		_ = pok
-
-		// TODO: implement this
-		// switch {
-		// case cok && hok:
-		// 	modifiedRw = closeNotifyHijackResponseWriter{gw}
-		// case cok && pok:
-		// 	modifiedRw = closeNotifyPusherResponseWriter{gw}
-		// case cok:
-		// 	modifiedRw = closeNotifyResponseWriter{gw}
-		// case hok:
-		// 	modifiedRw = hijackResponseWriter{gw}
-		// case pok:
-		// 	modifiedRw = pusherResponseWriter{gw}
-		// }
-
-		wrappedHandler(modifiedRw, r)
-	}
-}
-
-var bufferPool = &sync.Pool{
-	New: func() interface{} {
-		buf := make([]byte, 0, defaultMinSize)
-		return &buf
-	},
-}
+		defer grw.Close()
 
-var gzipWriterPools [gzip.BestCompression - gzip.HuffmanOnly + 1]sync.Pool
+		if _, ok := w.(http.CloseNotifier); ok {
+			// TODO: handle this case.
 
-func gzipWriterPool(level int) *sync.Pool {
-	return &gzipWriterPools[level-gzip.HuffmanOnly]
-}
+			// gwcn := GzipResponseWriterWithCloseNotify{gw}
+			// wrappedHandler(gwcn, r)
+			// return
+		}
 
-func gzipWriterGet(w io.Writer, level int) *gzip.Writer {
-	if gw, ok := gzipWriterPool(level).Get().(*gzip.Writer); ok {
-		gw.Reset(w)
-		return gw
+		wrappedHandler(grw, r)
 	}
-
-	gw, _ := gzip.NewWriterLevel(w, level)
-	return gw
-}
-
-func gzipWriterPut(gw *gzip.Writer, level int) {
-	gzipWriterPool(level).Put(gw)
 }
 
-// gzipRW provides an http.ResponseWriter interface,
-// which gzips bytes before writing them to the underlying
-// response. This doesn't close the writers, so don't forget
-// to do that. It can be configured to skip response smaller
-// than minSize.
+// gzipRW provides an http.ResponseWriter interface, which gzips
+// bytes before writing them to the underlying response. This doesn't close the
+// writers, so don't forget to do that.
+// It can be configured to skip response smaller than minSize.
 type gzipRW struct {
 	http.ResponseWriter
+	level int
```

Comment:
do not wrap responseWriter since it breaks optional interfaces 
https://github.com/rs/cors/pull/84#discussion_r294039093